### PR TITLE
chore(auth): Add `SignOutStateMachine`

### PR DESCRIPTION
**Pull request by dnys1**
Integrates the sign out flow into the state machine queueing system which helps prevent against some weird race conditions and ensure serializability of API calls.

### CODE REVIEW

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/sign_out_state_machine.dart [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+      if (hostedUiResult != null) {
+        return emit(SignOutState.complete(hostedUiResult));
+      }
+    }
+
+    // Do not try to send Cognito requests for plugin configs without an
+    // Identity Pool, since the requests will fail.
+    if (_identityPoolConfig != null) {
+      if (options.globalSignOut) {
+        // Revokes the refresh token
```
Comment (NikaHsn) [inline packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/sign_out_state_machine.dart:108]:
```suggestion
        // Revokes the access token
```

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/sign_out_state_machine.dart [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+          await _cognitoIdp
+              .globalSignOut(
+                GlobalSignOutRequest(
+                  accessToken: tokens.accessToken.raw,
+                ),
+              )
+              .result;
+        } on Exception catch (e) {
+          globalSignOutException = GlobalSignOutException(
+            accessToken: tokens.accessToken.raw,
```
Comment (NikaHsn) [inline packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/sign_out_state_machine.dart:119]: what is the need to capture accessToken.raw in this exception?

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/sign_out_state_machine.dart [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            underlyingException: e,
+          );
+        }
+      }
+      // Revokes the access token
```
Comment (NikaHsn) [inline packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/sign_out_state_machine.dart:124]:
```suggestion
      // Revokes the refresh token
```

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/sign_out_state_machine.dart [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+      if (globalSignOutException != null) {
+        revokeTokenException = RevokeTokenException(
+          refreshToken: tokens.refreshToken,
```
Comment (NikaHsn) [inline packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/sign_out_state_machine.dart:127]: what is the need to capture refreshToken in this exception?

Comment (dnys1) [reply packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/sign_out_state_machine.dart:127]: It gives users the opportunity to retry the revoke on their own. After this function completely, all the tokens are cleared locally.

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/sign_out_state_machine.dart (hunk 3 above)
Comment (dnys1) [reply packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/sign_out_state_machine.dart:124]: It's actually the reverse, confusingly. We revoke the access token with the refresh token and vice versa.

**Comment by NikaHsn** [comment]
Q: in sign_out_state_machine.dart we do SignOutState.complete even if the CognitoSignOutResult is failed or partially failed. why not to add failure state to SignOutStateType and use it for these cases?

**Comment by dnys1** [comment]
> Q: in sign_out_state_machine.dart we do SignOutState.complete even if the CognitoSignOutResult is failed or partially failed. why not to add failure state to SignOutStateType and use it for these cases?

Good point - I don't like that naming. I changed it to be more inline with other state machines and broke out the failure states.

**Review by ragingsquirrel3 — APPROVED** [review]

**Review by NikaHsn — APPROVED** [review]

**Pull request merged**